### ISSUE TITLE
status of hook_civicrm_alterUrl ?

### ISSUE THREAD

**Issue by JoeMurray** [WeMoveEU/utmaltor#4]
I notice in your documentation that
> This feature works based on not published hook hook_civicrm_alterUrl, yet.

I agree this hook is not present on https://docs.civicrm.org/dev/en/latest/hooks/list/

Is this hook in core? If not, have you submitted a PR to have it included?

**Comment by bhahumanists** [comment]
This would be a really useful hook! I can't see it in core or this extension. If it's available I'd be happy to help test it...

**Comment by scardinius** [comment]
Hi @JoeMurray, @bhahumanists,

here is an issue with PR about this hook https://lab.civicrm.org/dev/core/issues/1024
Even if it will not be merged  soon, it could be help to use for you.

**Comment by MegaphoneJon** [comment]
I think this can be closed since the answer is that [it won't be merged](https://github.com/civicrm/civicrm-core/pull/14462#issuecomment-560694919).  A Flexmailer-compatible alternative is necessary.

**Comment by JoeMurray** [comment]
Thanks, @scardinius but I too think this can be closed.

**Comment by MegaphoneJon** [comment]
@scardinius @JoeMurray @bhahumanists I'm working on a Flexmailer-compatible fork, which works without this hook: https://github.com/MegaphoneJon/utmaltor/tree/flexmailer

#### Mosaico
It seems to be working bug-free in Mosaico, or any future templating language built atop Flexmailer.  

#### Traditional templates
It's working for the typical mailing submission workflow, but there's a bug when sending test messages.

Flexmailer "abdicates" to `CRM_Mailing_BAO_MailingJob::deliver()` when using a traditional template - and `deliver()` pulls the data straight from the database.  So for traditional templates, I have to write the changes back to `civicrm_mailing`.  So the test mail goes out successfully - then CiviMail sees the changed db content, decides someone else edited the mailing since you last saved, and throws an error.  Reloading the page clears the error.I think I'm going to go with just reloading the page on sending a test mailing from the traditional template.

I also identified and fixed a couple of unrelated bugs that affect the current master branch.

**Comment by scardinius** [comment]
@MegaphoneJon  Thanks a lot for your modifications! Sure, I will test your branch

**Comment by scardinius** [comment]
@MegaphoneJon 

I tested **traditional** template 

Updating by SQL query (your commit) resolves problem on "Send test" action but not exactly. Why? Because next ajax call updates body_html to value from opened form.

There is a difference between utmaltor with custom alterUrl hook and utmaltor with flexmailer. Custom hook is called during composing a content of email, version with flexmailer alters urls during sending mailing.

**Comment by MegaphoneJon** [comment]
@scardinius Thank you for your review!

I agree that the behavior is different with traditional vs. non-abdicated Flexmailer (i.e. Mosaico), and that the SQL query is a workaround.  However, I think it's a decent workaround in that the user-facing behavior is consistent with using Mosaico.

I guess I'm wondering if there's a benefit (and how strong it is) to adding the UTMs at the time of composition vs. on send?

Also, I noticed that SQL error you picked up during review; I've been meaning to fix it, I'll do so now.  Thank you!

**Comment by scardinius** [comment]
@MegaphoneJon 

> I guess I'm wondering if there's a benefit (and how strong it is) to adding the UTMs at the time of composition vs. on send?

* 1st advantage: you know which url will be used in trackable url. Exactly the same is in html body, espacialy when base url has own params
* 2nd advantage: during composing content after reopen the form urls are altered and is easy to update by custom value. Also easier to maintenance valid format of url (non-technical users can get lost in ampersands, really)
* 3rd advantage: it's possible to use custom value instead of default from utmaltor setting page. utmaltor has configuration options like *Override existing utm_source params* which is TRUE as default setting. Change it to FALSE and you can use own value.

To sum up, your `flexmailer` branch is fine when above points are not important. In opposite situation, I need something different. And I can't use flexmailer events because their are too late, hmm. I don't know how civicrm-core will be updated

**Comment by MegaphoneJon** [comment]
@scardinius Hmm, I see what you're saying.  You're right, your use case needs more flexibility.  I also never used the original version of your extension; the first time I needed this, it was with a client using Flexmailer.

I'm going to take a deeper look at this.  My first thought is to look at using `hook_civicrm_pre` on the `Mailing` entity.  But I'm guessing you already investigated this.  Can you just let me know if you have, and if so, why it won't work?

**Comment by scardinius** [comment]
@MegaphoneJon Your suggestion to use `hook_civicrm_pre` would be excellent if we would talking only about links in content of mailing. However we're missing links placed in footer or header. Method `CRM_Mailing_BAO_Mailing->compose()` called before sending test or final mailing has all links, also those from footer and header. And alterUrl hook is placed there.

So, `hook_civicrm_pre` isn't enough :(

Edit: maybe the solution will be combine both approach, flexmailer as is now +  `hook_civicrm_pre`.

**Comment by scardinius** [comment]
@MegaphoneJon and a bit time later... things changed

I finally prepare solution based on civicrm hooks - post and alterMailContent with small patch on civicrm-core. Not necessary calls to any flexmailer code.

The patch is necessary to extend template used by alterMailContent. And what is a good news for us - there are a official the same/similar changes, see https://lab.civicrm.org/dev/core/issues/1620

So it's possible that next civicrm version (5.24?) will have all needed modification for utmaltor and then I will prepare release.
